docs/interactive_engine/tutorial_ldbc_gremlin.md

```diff
@@ -831,3 +831,1087 @@ At last, we would like to introduce `endV()` step in path expansion. Since a pat
+
+print(q1.all())
+print(q2.all())
+print(q3.all())
+print(q4.all())
+```
+
+The output should be like:
+
+```bash
+[v[216172782113784082], v[216172782113784169], v[216172782113784267], v[216172782113784368], ..., ]
+[v[216172782113784376], v[216172782113783938], v[216172782113784405], v[216172782113783980], ..., ]
+[v[216172782113783844], v[216172782113783901], v[216172782113783935], v[216172782113784439], ..., ]
+[v[216172782113784315], v[216172782113784374], v[216172782113784601], v[216172782113784431], ..., ]
+```
+
+### Limit Step
+
+Gremlin sentences always generate a lot of traversers, but sometimes we only need some of them. Therefore, gremlin provides `limit()` step to filter the objects in the traversal by the number of them to pass through the stream, where only the first n objects are allowed as defined by the limit argument. 
+
+For example, if you want to extract 10 person vertices from the Property Graph, you can write:
+
+```python
+# Extract 10 person vertices
+q1 = g.execute('g.V().hasLabel(\'person\').limit(10)')
+print(q1.all())
+```
+
+The output should be like:
+
+```bash
+[v[216172782113783808], v[216172782113783809], v[216172782113783810], v[216172782113783811], v[216172782113783812], v[216172782113783813], v[216172782113783814], v[216172782113783815], v[216172782113783816], v[216172782113783817]]
+```
+
+`limit()` step is usually used together with `order()` step to select *top(k)* vertices. For example, if you want to select 10 persons with the largest number of degrees, you can write in GIE as:
+
+```python
+# Extract 10 person vertices with the largest number of degrees
+q1 = g.execute('g.V().hasLabel(\'person\').order().by(both().count(), desc).limit(10)')
+print(q1.all())
+```
+
+The output should be like:
+
+```bash
+[v[216172782113784601], v[216172782113784315], v[216172782113784011], v[216172782113784374], v[216172782113783971], v[216172782113784431], v[216172782113784333], v[216172782113784154], v[216172782113784381], v[216172782113783933]]
+```
+
+## Complex Queries
+
+Finally, we will discuss how to use GIE with gremlin sentences to make some complex queries. The queries we choose are mainly from LDBC BI Workload.
+
+### LDBC BI2
+
+:::{figure-md}
+![bi2](../images/bi2.png)
+Figure 25. LDBC BI Query2.
+:::
+
+This figure illustrates LDBC BI2 query. This query aims to find the `Tags` under a given `TagClass` that were used in `Messages` during in the 100-day time window starting at date and compare it with the 100-day time window that follows. For the `Tags` and for both time windows, compute the count of `Messages`. 
+
+The key to this query is creating two sub execution branches to count the two windows separately. We can use gremlin's `by()` step to achieve the purpose. Assume the `TagClass's name='Actor'` `$date='2012-01-01'`, `$date+100='2012-04-09'$` and `$date+200='2012-07-18'`, we can write in GIE as:
+
+```python
+q1 = g.execute('g.V().has(\'tagclass\', \'name\', \'Actor\').in(\'hasType\').as(\'tag\')\
```

Comment:
I suggest breaking the BI 12 query step by step to actually instruct a user to how to program such a complex query.

Reply:
BTW. Just need to break down into one query, the others can be simply given for their reference.